src/Features/LanguageServer/Protocol/ExternalAccess/SimplifyTypeNamesHandler.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Simplify;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(SimplifyTypeNamesHandler)), Shared]
+[Method(SimplifyTypeNamesMethodName)]
+internal class SimplifyTypeNamesHandler : ILspServiceDocumentRequestHandler<SimplifyTypeNamesParams, string[]?>
+{
+    private const string SimplifyTypeNamesMethodName = "textDocument/simplifyTypeNames";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public SimplifyTypeNamesHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(SimplifyTypeNamesParams request) => request.TextDocument;
+
+    public async Task<string[]?> HandleRequestAsync(
+        SimplifyTypeNamesParams request,
+        RequestContext context,
+        CancellationToken cancellationToken)
+    {
+        var document = request.PlacementTextDocument is null
+            ? context.Document
+            : context.Solution!.GetDocument(request.PlacementTextDocument);
+
+        if (document is null)
+            return null;
+
+        var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        if (semanticModel is null)
```

Comment:
as this is C#/VB, this can't happen.  a better way to enforce that is simply `document.GetRequiredSemanticModelASync`.